system/security/LdapSecurity/ldapconnection.cpp

```diff
@@ -1121,30 +1121,40 @@ class CLdapClient : public CInterface, implements ILdapClient
     virtual void init(IPermissionProcessor* pp)
     {
         m_pp = pp;
-        if(m_ldapconfig->getServerType() == OPEN_LDAP)
+        static CriticalSection  lcCrit;
+        static bool createdOU = false;
+        if (!createdOU)
         {
-            try
-            {
-                addDC(m_ldapconfig->getBasedn());
-            }
-            catch(...)
-            {
-            }
-            try
-            {
-                addGroup("Directory Administrators", m_ldapconfig->getBasedn());
-            }
-            catch(...)
+            CriticalBlock block(lcCrit);
+            if (!createdOU)
```

Comment:
I am confused about your comment "the createdOU inside the critical block does not need to be re-evaluated".  That second test is for threads that were blocked by the CS while another thread eventually releases it after setting createdOU to true.  So without reevaluating, they would enter that block of code that doesn't need to be reexecuted.